### DISCUSSION TITLE
Idea: enable customization of ACI RGB values for drawing add-on export

### DISCUSSION THREAD

**Discussion by seanpk** [mozman/ezdxf#1051] [Ideas]
Simply provide some greater control over the colors used to render ACI colors from the drawing add-on.

It is useful to adjust the colors to have better contrast on some background (it is already possible to set the background color and the contrast color with `LayoutProperties`). It could also be useful to force all elements to a single color, say blue, to provide a simplified visual, say like a blueprint.

Just "reaching into" the `ezdxf.colors` module and changing the color values in the `DXF_DEFAULT_COLORS` and `DXF_DEFAULT_PAPERSPACE_COLORS` lists doesn't seem to have the intended effect.

I could work on a PR, assuming this would be a welcome feature.

**Comment by seanpk** [comment]
hmm. I just noticed this comment at the top of the `draw_layout()` implementation in `frontend.py`, is it related to how it seemed that setting values in those lists didn't work:
```python
        if layout_properties is not None:
            # TODO: this does not work, layer properties have to be re-evaluated!
            self.ctx.current_layout_properties = layout_properties
        else:
            self.ctx.set_current_layout(layout)
```

**Reply by mozman** [reply]
This seems to be an old comment, the feature works as expected in the `qsave()` function of the `Matplotlib` backend.

This is an old feature, it is better to set the layout properties by the config settings  `color_policy` and `background_policy`.

**Reply by seanpk** [reply]
ok, that is good, I will check with those

**Comment by mozman** [comment]
The [ColorPolicy](https://ezdxf.mozman.at/docs/addons/drawing.html#colorpolicy) does what you want.

```python
# black and white output
config = Configuration(
    color_policy = ColorPolicy.BLACK,
    background_policy=BackgroundPolicy.WHITE,  
)

# or map all colors to custom color
config = Configuration(
    color_policy = ColorPolicy.CUSTOM,
    custom_fg_color = "#FF0000",
    background_poliy=BackgroundPolicy.CUSTOM,
    custom_bg_color = "#0000FF",
)

frontend = Frontend(
    ctx=RenderContext(doc), 
    out=svg.SVGBackend(), 
    config=config,
)
```

## Additional ways to override the ACI

The way CAD applications allow users to set their own ACI colors is the `.ctb` plot style file and they have GUI applications to manage these `.ctb` files. `Ezdxf` can load and create `.ctb` files via the `ezdxf.addons.acadctb` [add-on](https://ezdxf.mozman.at/docs/addons/acadctb.html).

The `.ctb` file is evaluated by the `RenderContext` class, and this class is also the point where to override the associated `.ctb` file stored in the LAYOUT entity. `Ezdxf` searches the support directories defined in the config file for the `.ctb` files.

Override `.ctb` files:
```python
   ctx = RenderContext(doc, ctb="your.ctb")
```

Another way to override the default ACI colors is to inherit from the `Frontend` class and override the `override_properties()` method:

```python

class MyFrontend(Frontend):
    def override_properties(self, entity: DXFGraphic, properties: Properties) -> None:
        if properties.pen = 7:  # the ACI
            properties.color = "#AABBCC"
```

These are old features and there are better ways to do that.  The `override_properties()` method should be a function to pass in and not a method to override. The `.ctb` file should be set as an `acadctb.PlotStyle` instance, so you don't have to load it from the file system and you can create a new one on the fly in the script.  

But the `drawing` add-on is complex and I don't want to break existing code.

EDIT: I can/will fix this two problems without breaking existing code.

**Reply by seanpk** [reply]
What a great answer! I will make use of the capabilities that exist now and look forward to those two fixes/features becoming available. If there is anything I can do to assist, let me know :-)

Thank you!
